18-websockets/classworks/Malkov_Aleksei/src/pages/Chat/Chat.jsx

```diff
@@ -23,34 +23,40 @@ function Chat() {
 
   useEffect(() => {
     socket.emit("add user", defaultName);
+
     return () => {
       socket.disconnect();
     }
+    // eslint-disable-next-line
   }, []);
 
-  socket.on('reg as user', (data) => {
-    setUsername(data);
-  });
+  useEffect(() => {
+    socket.on('reg as user', (data) => {
+      setUsername(data);
+    });
+  }, [username]);
 
-  socket.on('new message', (data) => {
-    setListOfMessages([...listOfMessages, data]);
-  });
+  useEffect(() => {
+    socket.on('new message', (data) => {
+      setListOfMessages([...listOfMessages, data]);
+    });
+  }, [listOfMessages]);
 
-  socket.on('typing', (data) => {
-    if (!typing.includes(data.username)) {
-      setTimeout(() => {
-        setTyping([...typing].filter((typer) => typer !== data.username));
-      }, 2000)
-      setTyping([...typing, data.username]);
-    }
-  });
+  useEffect(() => {
+    socket.on('typing', (data) => {
+      //
+      if (!typing.includes(data.username)) {
+        setTimeout(() => {
+          setTyping([...typing.filter((typer) => typer !== data.username)]);
+        }, 2000)
+        setTyping([...typing, data.username]);
```

Comment:
Looks like there is one problem. You don't update user typing info if user has been already typing. For example, if user types and types again in one sec, other users will see typing message only for 2 seconds, not for 3.
The best way to fix this, it to reset the timer every time when you got an event about typing.